### PR TITLE
ONNX Object Detection Tutorial

### PR DESCRIPTION
**Pull request by luisquintanilla**
Object Detection Tutorial using ONNX

[AB#1450817](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1450817)

[AB#1566043](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1566043)

[Internal Review Link](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/object-detection-onnx?branch=pr-en-us-12755)

To-Do:

- [x] Add inspecting with Netron section
- [x] Add conceptual information (DNN, YOLO, ONNX)
- [x] Improve sections / headers
- [x] Add xrefs
- [x] Replace Hard-Coded Code Snippets with Machine Learning Samples

### CODE REVIEW

**Review by JRAlexander — CHANGES_REQUESTED** [review]
Looks good. Just a few minor changes.

**Review by JRAlexander — APPROVED** [review]

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 1]
```diff
@@ -0,0 +1,1124 @@
...
+
+    Console.WriteLine("");
+}
+```
+
+Both of these methods will be useful when the model has produced outputs and those have been processed. First though, create the functionality to process the model outputs.
+
+## Create a parser to post-process model outputs
+
+The expected output generated by the pre-trained ONNX model is a tensor of shape `125 x 13 x 13` that is returned as a one-dimensional `float[]` of 21125 elements. In order to extract the predictions generated by the model, some post-processing work is required. To do so, create a set of classes to help parse the output.
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: Should we define what a tensor is?

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 2]
```diff
@@ -0,0 +1,1124 @@
...
+
+Object detection is an image processing task. Therefore, most deep learning models trained to solve this problem are CNNs. The model used in this tutorial is the Tiny YOLOv2 model, a more compact version of the YOLOv2 model described in the paper: ["YOLO9000: Better, Faster, Stronger" by Redmon and Fadhari](https://arxiv.org/pdf/1612.08242.pdf). Tiny YOLOv2 is trained on the Pascal VOC dataset and is made up of 15 layers that can predict 20 different classes of objects. Because Tiny YOLOv2 is a condensed version of the original YOLOv2 model, a tradeoff is made between speed and accuracy. 
+
+The model is stored in ONNX format. The Open Neural Network Exchange (ONNX) is an open source format for AI models. ONNX supports interoperability between frameworks. This means you can train a model in one of the many popular machine learning frameworks like PyTorch, convert it into ONNX format and consume the ONNX model in a different framework like ML.NET. To learn more, visit the [ONNX website](https://onnx.ai/). 
+
+![](./media/object-detection-onnx/onnx-frameworks.png)
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: A diagram with the network inputs and outputs would be very useful here. If you draft it, we can ask Deborah in art to make it professional

File: docs/machine-learning/tutorials/object-detection-onnx.md (hunk 1 above)
Comment (natke) [reply docs/machine-learning/tutorials/object-detection-onnx.md]: A diagram showing the mapping would be useful here

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 3]
```diff
@@ -0,0 +1,1124 @@
...
+
+# Tutorial: Detect objects using ONNX in ML.NET
+
+Learn how to use a pre-trained ONNX model in ML.NET to detect objects. The pre-trained model was trained to find and classify objects in an image.
+
+Training an object detection model from scratch requires setting millions of parameters, a large amount of labeled training data and a vast amount of compute resources (hundreds of GPU hours). Using a pre-trained model allows you to shortcut the training process.
+
+In this tutorial, you learn how to:
+> [!div class="checklist"]
+> * Understand the problem
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: I think we could expand this list to make it a little less generic. They learn a lot in this tutorial!

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 4]
```diff
@@ -0,0 +1,1124 @@
...
+
+Object detection is an image processing task. Therefore, most deep learning models trained to solve this problem are CNNs. The model used in this tutorial is the Tiny YOLOv2 model, a more compact version of the YOLOv2 model described in the paper: ["YOLO9000: Better, Faster, Stronger" by Redmon and Fadhari](https://arxiv.org/pdf/1612.08242.pdf). Tiny YOLOv2 is trained on the Pascal VOC dataset and is made up of 15 layers that can predict 20 different classes of objects. Because Tiny YOLOv2 is a condensed version of the original YOLOv2 model, a tradeoff is made between speed and accuracy. 
+
+The model is stored in ONNX format. The Open Neural Network Exchange (ONNX) is an open source format for AI models. ONNX supports interoperability between frameworks. This means you can train a model in one of the many popular machine learning frameworks like PyTorch, convert it into ONNX format and consume the ONNX model in a different framework like ML.NET. To learn more, visit the [ONNX website](https://onnx.ai/). 
+
+![](./media/object-detection-onnx/onnx-frameworks.png)
+
+## Create console application
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: Is this the best title for this section? Can we make it more specific?

**Review by natke — COMMENTED** [review]
Looks good! I've done an initial review, by just reading. I want to run through it again, and actually do the tutorial, and will probably add more comments then.

Oh, have we run acrolinx over this?

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 5]
```diff
@@ -0,0 +1,1124 @@
...
+    public class ImageNetPrediction
+    {
+        [ColumnName("grid")]
+        public float[] PredictedLabels;
+    }
+    ```
+
+    `ImageNetPrediction` is the prediction data class and has the following `float[]` field:
+
+    - `PredictedLabel` contains the dimensions, objectness score and class probabilities for each of the bounding boxes detected in an image.
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/object-detection-onnx.md]: That's what it's referred to as in the paper.

https://arxiv.org/pdf/1612.08242.pdf

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+ms.author: luquinta
+ms.date: 07/29/2019
+ms.topic: tutorial
+ms.custom: mvc
+#Customer intent: As a developer, I want to use ML.NET so that I can use a pre-trained model in an object detection scenario to detect objects in images using ONNX.
+---
+
+# Tutorial: Detect objects using ONNX in ML.NET
+
+Learn how to use a pre-trained ONNX model in ML.NET to detect objects. The pre-trained model was trained to find and classify objects in an image.
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: Suggest simplifying this to: 
```
Learn how to use a pre-trained ONNX model in ML.NET to detect objects in images
```

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+- [Tiny YOLOv2 pre-trained model](https://github.com/onnx/models/tree/master/tiny_yolov2)
+- [Netron](https://github.com/lutzroeder/netron) (optional)
+
+## ONNX object detection sample overview
+
+This sample creates a .NET core console application that detects objects within an image using a pre-trained deep learning ONNX model. The code for this sample can be found on the [dotnet/machinelearning-samples repository](https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/getting-started/DeepLearning_ObjectDetection_Onnx) on GitHub.
+
+## What is object detection?
+
+Object detection is a computer vision problem. While closely related to image classification, object detection performs image classification at a more granular scale. When an image is processed, there's first a check on whether there's an object present or not. If so, the detected object is classified and using object localization, the location of that object is returned along with the class. Typically, you want to use this approach on images where there are multiple objects of different types so that each is individually recognized.
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: Suggest simplifying to:
`Object detection both locates _and_ categorizes entities within images. Use object detection when images contain multiple objects of different types.`

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 8]
```diff
@@ -0,0 +1,1124 @@
...
+Some use cases for object detection include:
+
+- Self-Driving Cars
+- Robotics
+- Face Detection
+- Workplace Safety
+- Object Counting
+- Activity Recognition
+
+## Select the appropriate machine learning task
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/tutorials/object-detection-onnx.md]: I'm still of the opinion that this title doesn't describe the content of this section

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+
+There are different types of neural networks, each more suitable than the other for certain tasks. The most common types of neural networks are Multi-Layered Perceptron (MLP), Convolutional Neural Network (CNN) and Recurrent Neural Network (RNN). The most basic is the MLP, which maps a set of inputs to a set of outputs. This neural network is good when the data does not have a spatial or time component. The CNN as implied by the name makes use of convolutional layers that take into account spatial information contained in the data. A good use case for CNNs is image processing when trying to detect the presence of a particular feature in a region of an image (i.e. is there a nose in the center of an image?). Finally, RNNs allow for the persistence of state or memory to be used as input therefore making it a good candidate for time-series analysis where the sequential ordering and context of events is important.
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]:
```suggestion
There are different types of neural networks, the most common being Multi-Layered Perceptron (MLP), Convolutional Neural Network (CNN) and Recurrent Neural Network (RNN). The most basic is the MLP, which maps a set of inputs to a set of outputs. This neural network is good when the data does not have a spatial or time component. The CNN makes use of convolutional layers to process spatial information contained in the data. A good use case for CNNs is image processing to detect the presence of a feature in a region of an image (for example, is there a nose in the center of an image?). Finally, RNNs allow for the persistence of state or memory to be used as input. RNNs are used for time-series analysis, where the sequential ordering and context of events is important. 
```

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+
+![](./media/object-detection-onnx/netron-model-map.png)
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: The netron part of this image is too small. I can help rework this image

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+
+The input expected by the YOLO model is an image that is `416px x 416px`, where each pixel contains 3 data elements corresponding to the RGB values of the respective pixel. The model takes this input and passes it through the different layers to produce an output. The output is a grid of dimensions `125 x 13 x 13`.
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]:
```suggestion
The YOLO model takes an image `416px x 416px x 3 (RGB)`. The model takes this input and passes it through the different layers to produce an output. The output divides the input image into a `13 x 13` grid, with each box in the grid consisting of `125` values. 
```

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+
+![](./media/object-detection-onnx/netron-model-map.png)
+
+The input expected by the YOLO model is an image that is `416px x 416px`, where each pixel contains 3 data elements corresponding to the RGB values of the respective pixel. The model takes this input and passes it through the different layers to produce an output. The output is a grid of dimensions `125 x 13 x 13`. 
+
+![](./media/object-detection-onnx/model-output-description.png)
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: Image looks good! Would the yellow bounding box be where the centroid of the object is though?

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+
+The model segments an image into a `13 x 13` grid, where each grid cell is `32px x 32px`. Each grid cell contains 125 elements within it that correspond to 5 areas in an image which are believed to contain an object, also known as bounding boxes. Each of these bounding boxes stores 25 elements of information to describe itself:
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]:
```suggestion
The model segments an image into a `13 x 13` grid, where each grid cell is `32px x 32px`. Each grid cell contains 5 potential object bounding boxes. A bounding box has  25 elements:
```

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 14]
```diff
@@ -0,0 +1,713 @@
...
+
+    The *ImageNetPrediction.cs* file opens in the code editor. Add the following `using` statement to the top of *ImageNetPrediction.cs*:
+
+    [!code-csharp [ImageNetPredictionUsings](~/machinelearning-samples/samples/csharp/getting-started/DeepLearning_ObjectDetection_Onnx/ObjectDetectionConsoleApp/DataStructures/ImageNetPrediction.cs#L1)]
+
+    Remove the existing class definition and add the following code for the `ImageNetPrediction` class to the *ImageNetPrediction.cs* file:
+
+    [!code-csharp [ImageNetPredictionClass](~/machinelearning-samples/samples/csharp/getting-started/DeepLearning_ObjectDetection_Onnx/ObjectDetectionConsoleApp/DataStructures/ImageNetPrediction.cs#L5-L9)]
+
+    `ImageNetPrediction` is the prediction data class and has the following `float[]` field:
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]: I suggest moving the grid image into this section, to break up the code a bit, and also to illustrate how the ImagePrediction data is encoded.

We could talk about the expected dimensions of the ImagePrediction float[] array too

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 15]
```diff
@@ -0,0 +1,713 @@
...
+
+The output generated by the pre-trained ONNX model is a float array of length 21125, representing the elements of a tensor with dimensions `125 x 13 x 13`. In order to transform the predictions generated by the model into a tensor, some post-processing work is required. To do so, create a set of classes to help parse the output.
+
+Add a new directory to your project to organize the set of parser classes.
+
+1. In **Solution Explorer**, right-click the project, and then select **Add** > **New Folder**. When the new folder appears in the Solution Explorer, name it "YoloParser".
+
+### Create bounding boxes and dimensions
+
+Object localization is a core component of object detection. As such, much of the data output by the model contains coordinates and dimensions of the bounding boxes of objects within the image. Create a base class for dimensions.
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]:
```suggestion
The data output by the model contains coordinates and dimensions of the bounding boxes of objects within the image. Create a base class for dimensions.
```

**Review by natke — COMMENTED** [review]
Overall, looking good! I have a few more suggestions.

File: docs/machine-learning/tutorials/object-detection-onnx.md [hunk 16]
```diff
@@ -0,0 +1,713 @@
...
+
+![](./media/object-detection-onnx/netron-model-map.png)
+
+The YOLO model takes an image `3(RGB) x 416px x 416px`. The model takes this input and passes it through the different layers to produce an output. The output divides the input image into a `13 x 13` grid, with each box in the grid consisting of `125` values.
```
Comment (natke) [inline docs/machine-learning/tutorials/object-detection-onnx.md]:
```suggestion
The YOLO model takes an image `3(RGB) x 416px x 416px`. The model takes this input and passes it through the different layers to produce an output. The output divides the input image into a `13 x 13` grid, with each cell in the grid consisting of `125` values. 
```

**Review by natke — COMMENTED** [review]
One small comment, which was actually my bad from the last review: box --> cell

**Review by natke — APPROVED** [review]
